x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts

```diff
@@ -324,3 +324,13 @@ export const internalRuleUpdate = t.intersection([
   internalRuleUpdateRequired,
 ]);
 export type InternalRuleUpdate = t.TypeOf<typeof internalRuleUpdate>;
+
+// Rule params of rules with index and dataViewId properties
+export type IndexPatternRuleParams =
+  | EqlRuleParams
+  | ThreatRuleParams
+  | QueryRuleParams
+  | QueryRuleParams
+  | SavedQueryRuleParams
+  | ThresholdRuleParams
+  | NewTermsRuleParams;
```

Comment:
Nit: I wonder if this type really helps. The `index` property is still optional...

Reply:
I think I need a little helps with the types here. I see what you mean about `index` (and also `dataViewId`!) still being optional here, but for some reason this union type allows me to type the `isDataViewExistsAndNotOverriden`, `hasIndexPatterns` and `hasNotIndexPattern` correctly. Otherwise, if I just set the type to `RuleAlertType['params']` TS will complain that both `index` and `dataViewId` can be undefined.

Any ideas why?

Reply:
@jpdjere here's some suggestion:

```ts
// Check if current params have a configured data view id
// and the action is not set to overwrite data views
const isDataViewExistsAndNotOverriden = (
  dataViewId: string | undefined,
  action: BulkActionEditPayloadIndexPatterns
) => dataViewId != null && !action.overwrite_data_views;

// Check if the index patterns added to the rule already exist in it
const hasIndexPatterns = (
  indexPatterns: string[] | undefined,
  action: BulkActionEditPayloadIndexPatterns
) => action.value.every((indexPattern) => indexPatterns?.includes(indexPattern));

// Check if the index patterns to be deleted don't exist in the rule
const hasNotIndexPattern = (
  indexPatterns: string[] | undefined,
  action: BulkActionEditPayloadIndexPatterns
) => action.value.every((indexPattern) => !indexPatterns?.includes(indexPattern));

const shouldSkipIndexPatternsBulkAction = (
  indexPatterns: string[] | undefined,
  dataViewId: string | undefined,
  action: BulkActionEditPayloadIndexPatterns
) => {
  if (isDataViewExistsAndNotOverriden(dataViewId, action)) {
    return true;
  }

  if (action.type === BulkActionEditType.add_index_patterns) {
    return hasIndexPatterns(indexPatterns, action);
  }

  if (action.type === BulkActionEditType.delete_index_patterns) {
    return hasNotIndexPattern(indexPatterns, action);
  }

  return false;
};
```

```ts
      if (shouldSkipIndexPatternsBulkAction(ruleParams.index, ruleParams.dataViewId, action)) {
        isActionSkipped = true;
        break;
      }
```

After that we could delete the `IndexPatternRuleParams`.

Reply:
Thanks, definitely simpler